TeamCity CodeStyle: lint the modified files in batches

oandregal opened the pull request:
## Proposed Changes

For linting the files modified, we have to cover for two issues:

- ENAMETOOLONG: we cannot pass to the ESLint process too many files as input because we can exceed the maximum command line length.
- OOM: we cannot spawn too many processes in parallel, otherwise TeamCity may run Out Of Memory.

To prevent running into any of these issues, this PR processes the files in batches:

- `BATCH_SIZE` is the number of files each ESLint process will take.
- `MAX_PARALLEL_BATCHES` is the maximum number of ESLint processes to execute in parallel.

## Why are these changes being made?

To cover for the above issues, right now we use the following strategy:

1. If there are more than 16 files to lint, lint the whole codebase.

This prevents both ENAMETOOLONG and OOM issues. It comes at the cost of impacting big PRs by requiring them to lint files they didn't modify. This is opaque to the PR authors (they need to dig into this job to understand) and creates friction.

2. If there are less than 16 files to lint, spawn a ESLint process for each file.

This takes advantage of parallelism, but only in a limited way (ESLint only process 1 file per process). It aims to control ENAMETOOLONG by only linting 1 file, and OOM by only spawning 16 processes.

By spawning `MAX_PARALLEL_BATCHES` (15) we take advantage of parallelism, and by limiting each ESLint process to only lint `BATCH_SIZE` (15) we cover against ENAMETOOLONG issues. These numbers can be tweaked.

## Tests

- [x] 0 files to lint (this PR as it is). Tests linting the whole codebase.
- [x] [101074](https://github.com/Automattic/wp-calypso/pull/101074). Tests that it will only lint the modified files, and that it does it in batches.
- [x] [101077](https://github.com/Automattic/wp-calypso/pull/101077) Tests that it won't spawn more than `MAX_PARALLEL_BATCHES` batches.

tyxla commented on .teamcity/_self/projects/WebApp.kt:
```diff
@@ -579,7 +579,7 @@ object CheckCodeStyleBranch : BuildType({
 				if [ "%run_full_eslint%" = "true" ] || [ "${'$'}TOTAL_FILES_TO_LINT" -gt 16 ] || [ "${'$'}TOTAL_FILES_TO_LINT" == "0" ]; then
 					echo "Linting all files"
-					yarn run eslint --format checkstyle --output-file "./checkstyle_results/eslint/results.xml" .
+					yarn run eslint --format checkstyle --output-file "./checkstyle_results/eslint/results.xml" ${'$'}(git diff --name-only --diff-filter=d refs/remotes/origin/trunk...HEAD | grep -E '(\.[jt]sx?|\.json|\.md)${'$'}'
```
Seems like this will fail if zero files are found to be linted:

![Screenshot 2025-03-07 at 11 23 26](https://github.com/user-attachments/assets/03a209d8-610a-4d11-8c4a-0ca191fb1b13)

oandregal replied:
Updated the approach to take into account other factors. See updated PR description.

oandregal committed:
Lint modified files in batches

oandregal commented on .teamcity/_self/projects/WebApp.kt:
```diff
@@ -575,19 +575,61 @@ object CheckCodeStyleBranch : BuildType({
+
+						if [ -n "${'$'}BATCH_FILES" ]; then
+							echo "Linting batch ${'$'}BATCH_NUM of ${'$'}TOTAL_BATCHES"
+							yarn run eslint --format checkstyle --output-file "./checkstyle_results/eslint/batch_${'$'}{BATCH_NUM}.xml" ${'$'}BATCH_FILES &
```
This is going to generate a different file than the existing config. Need to figure out how this file is being used in the next steps.

oandregal commented:
@aidvu @tyxla @jsnajdr The three tests (this PR, https://github.com/Automattic/wp-calypso/pull/101074, and https://github.com/Automattic/wp-calypso/pull/101077) seem to be working as expected. I'd welcome a review here.

tyxla commented:
With the changes you introduced, I wonder if it still makes sense to run a full lint if we have zero files to lint. What are we aiming to do? 

I also wonder if we should run a full lint for better precision in a few other scenarios:
* When an ESLint config file is changed
* When TC config is changed
* When there are ESLint package changes

oandregal commented:
> With the changes you introduced, I wonder if it still makes sense to run a full lint if we have zero files to lint. What are we aiming to do?

I think that's a good conversation to have, and that can be implemented separately. I'd like to keep this PR focused on the issue (don't trigger a full-codebase lint when the PR modifies a few files).

tyxla commented on .teamcity/_self/projects/WebApp.kt:
```diff
@@ -575,19 +575,61 @@ object CheckCodeStyleBranch : BuildType({
 
 				# Find files to lint
 				TOTAL_FILES_TO_LINT=$(git diff --name-only --diff-filter=d refs/remotes/origin/trunk...HEAD | grep -cE '(\.[jt]sx?|\.json|\.md)${'$'}' || true)
+				FILES_TO_LINT=$(git diff --name-only --diff-filter=d refs/remotes/origin/trunk...HEAD | grep -E '(\.[jt]sx?|\.json|\.md)${'$'}' || echo "")
```
Can we calculate `TOTAL_FILES_TO_LINT` out of `FILES_TO_LINT` to avoid repetition? Perhaps by using `wc -l` or something similar?

mcsf replied:
Two-part question:

* I don't see Bash's  `set -e` option in this script, i.e. there is no explicit setting to terminate execution at the first error. Is there some context that I need to know (e.g. all scripts run through TeamCity implicitly have `set -e`)? If there is no such behaviour in place, then the `|| echo ""` alternative is not needed. Or... we decide to add `set -e`, which is probably good discipline.
* If we do need to catch errors, then we should use the cleaner (and idiomatic) `|| true` instead of `echo`, which is noisy.

The first question also applies to `TOTAL_FILES_TO_LINT`.

I'd also improve the variable naming, because it wasn't obvious (the hint is the extra `c` parameter passed to `grep`) that one is a count and the other is a list of filenames.

oandregal replied:
> I don't see Bash's set -e option in this script

We now know the answer to this question. TeamCity sets this up and that's why https://github.com/Automattic/wp-calypso/pull/100986/commits/bd6103dc87f5e1522b1bfdf891d00fcd69d58862

tyxla reviewed (approved):
This works well! Nice work with the batching 🚀 

Let's keep an eye on the code style jobs for a few days after this lands.

Thanks @oandregal!

oandregal commented:
Thanks for the review, Marin!

I was discussing this with @mcsf and he mentioned this could leverage [xargs for batching](https://lamda.blog/2024/12/26/batching-tasks-with-xargs/). I'd like to wait for his full review before merging, in case he comes up with other things to improve.

mcsf commented on .teamcity/_self/projects/WebApp.kt:
```diff
@@ -575,19 +575,61 @@ object CheckCodeStyleBranch : BuildType({
+				# Create temporary output directory
+				mkdir -p "./checkstyle_results/eslint"
+
+				if [ "%run_full_eslint%" = "true" ] || [ "${'$'}TOTAL_FILES_TO_LINT" == "0" ]; then
```
I'm erring on the side of caution here, because `== "0"` actually behaves the same way. The reason why I'm suggesting this change is because it could lead novice Bash users to think that number comparisons are possible with `== != > <` etc., whereas `-eq -ne -gt -lt` should be used.

```suggestion
				if [ "%run_full_eslint%" = true ] || [ "${'$'}TOTAL_FILES_TO_LINT" -eq 0 ]; then
```

mcsf replied:
Second question: I'm having a hard time finding a good reference on the syntax expected by TeamCity inside `scriptContent`. Although I don't like asking LLMs for this sort of information, one model suggested that dollar signs can be escaped via simple repetition: `$$TOTAL_FILES_TO_LINT`. This wouldn't surprise me, since it's a convention that Makefiles started 48 years ago. :)

If that's the case and not hallucination, we should clean up this whole script.

oandregal replied:
> Second question: I'm having a hard time finding a good reference on the syntax expected by TeamCity inside scriptContent

We know the answer to this now :) https://github.com/Automattic/wp-calypso/pull/100986#issuecomment-2714913974

mcsf reviewed (commented):
Left some questions. I'll play with this a bit in the meantime. :)

mcsf committed:
Rewrite batched linting with xargs

mcsf commented:
I pushed a rewrite of the batching using `xargs` (and a couple of tricks). I commented the logic _in situ_ abundantly.

One thing that remains to be tested is whether `$$` is adequate escaping of `$`. 🤞

tyxla commented:
> One thing that remains to be tested is whether `$$` is adequate escaping of `$`. 🤞

You might want to try with `${'$'}`

mcsf commented:
> > One thing that remains to be tested is whether `$$` is adequate escaping of `$`. 🤞
> 
> You might want to try with `${'$'}`

Yes, but I raised the [question here](https://github.com/Automattic/wp-calypso/pull/100986#discussion_r1989281493) that, lacking good information on escaping rules, I found something suggesting that `$$` would work (which would be more legible than `${'$'}`). If it fails, we'll use what we had before.

mcsf committed:
Fix syntax: $$ -> ${'$'}

mcsf committed:
Restore 'set -x'

mcsf commented:
> > > One thing that remains to be tested is whether `$$` is adequate escaping of `$`. 🤞
> > 
> > 
> > You might want to try with `${'$'}`
> 
> Yes, but I raised the [question here](https://github.com/Automattic/wp-calypso/pull/100986#discussion_r1989281493) that, lacking good information on escaping rules, I found something suggesting that `$$` would work (which would be more legible than `${'$'}`). If it fails, we'll use what we had before.

André just reached out to me: this syntax did fail (it's to be used for compile-time references, it seems), so I've moved everything to `${'$'}`.

mcsf commented:
> > André just reached out to me: this syntax did fail (it's to be used for compile-time references, it seems), so I've moved everything to `${'$'}`.
> 
> Yep, TC is odd. And I was just trying to save you a few minutes, since we've fought with that before 😅

I guess this idiot had to make the mistake for himself. 😄

mcsf committed:
Readd `|| true` constructs

It turns out that, yes, TeamCity implicitly runs with `set -e`

oandregal commented:
Rebased the test PRs using the new approach, now that TeamCity has reported that it works fine.

oandregal commented:
@mcsf Looking at the logs generated by TeamCity for the two tests PRs ([one](https://teamcity.a8c.com/buildConfiguration/calypso_calypso_WebApp_Check_Code_Style_Branch/14485475?buildTab=log&linesState=499&logView=flowAware), [two](https://teamcity.a8c.com/buildConfiguration/calypso_calypso_WebApp_Check_Code_Style_Branch/14485494?buildTab=log&linesState=514&logView=flowAware)), it looks like the `rs` command wasn't found.

Judging by your comment:

> where -n3 is the batch size and -P5 the number of parallel runs. However, we
> want to know which batch we are currently in -- a concept that I don't think
> xargs has.

I understand you've introduced this to be able to output progress. I consider that a nice to have that I introduced in the previous implementation (`for` loop) because it was convenient to see that it worked. If we use xargs batching mechanism, I don't think we need to "make sure it works", and can remove it — some thoughts in case it's complicated to have.

oandregal commented on .teamcity/_self/projects/WebApp.kt:
```diff
@@ -571,23 +571,67 @@ object CheckCodeStyleBranch : BuildType({
-					wait
+					echo "Linting affected files"
+
+					# In an ideal scenario, we'd simply pipe the list of target files to `xargs`:
```
It'd be great to add a comment explaining _why_ we batch (to prevent both OOM and ENAMETOOLONG errors). That's a piece of knowledge that future readers (including ourselves) will only get if they dig into git history and find this PR.

mcsf replied:
You're right; I'll put it back.

mcsf committed:
Replace `rs` (not POSIX) with `awk`

TIL `rs` is not POSIX, so we can't rely on that. `nl` is fine, though.
The AWK script is less legible than `rs`, but its difficulty is
contained and IMO doesn't harm the rest of the sequence too much.
Hopefully the comments will guide the reader.

mcsf committed:
Reinstate comment about ENAMETOOLONG and OOM errors

oandregal commented:
I saw that the Code Style job hadn't run, so I re-triggered naively. Looking now more deeply, I see that there's a deeper issue I don't understand:

The build is waiting for an agent to be available (it says there's no idle agents):

<img width="502" alt="Screenshot 2025-03-12 at 08 51 54" src="https://github.com/user-attachments/assets/a6936ba1-1a59-4cb2-acbd-8ea6e5373198" />

However, there are **FIVE** idle agents:

<img width="474" alt="Screenshot 2025-03-12 at 08 52 08" src="https://github.com/user-attachments/assets/30caccda-f746-433a-9edc-83660824d5db" />

oandregal committed:
Escape the % sign because Kotlin DSL interprets it as a parameter reference

oandregal commented:
All the agents were reported as "incompatible" (see "Build > Compatible agents" tab). It looks like the use of `%` in the latest changes at https://github.com/Automattic/wp-calypso/pull/100986/commits/834e9fca61b8ce6da3d1d7d5d1ef0915bc37b8ba was something Kotlin DSL interpreted as a parameter — so I escaped it at https://github.com/Automattic/wp-calypso/pull/100986/commits/d73aa71c988f2375e56a8ef56a8a725e74acd9a7. It's building now.

mcsf commented:
> All the agents were reported as "incompatible" (see "Build > Compatible agents" tab). It looks like the use of `%` in the latest changes at [834e9fc](https://github.com/Automattic/wp-calypso/commit/834e9fca61b8ce6da3d1d7d5d1ef0915bc37b8ba) was something Kotlin DSL interpreted as a parameter — so I escaped it at [d73aa71](https://github.com/Automattic/wp-calypso/commit/d73aa71c988f2375e56a8ef56a8a725e74acd9a7). It's building now.

Thanks very much! 🤞

mcsf committed:
Dismiss ESLint's exit code 1, propagate other errors

xargs will return 123 if any run returns a non-zero value. Let's ensure
we only catch relevant issues. ESLint's exit codes seem to be:

- 0 for no errors
- 1 for linting errors (let's ignore)
- 2 for other errors (let's propagate)

mcsf commented:
André and I talked: things seem to be working now, but the test PR builds both reveal a `123` exit code. What happens is:

- ESLint runs return `1` when there are linting errors
- xargs returns a non-zero code if any of its runs returned non-zero. In some implementations, xargs returns `123`, and in others `1`. Note that it still performed **all** the runs.

I pushed a commit (15dc2b1) to explicitly handle return codes:

- When an ESLint run returns `0`, there were no errors at all. Nothing to do.
- When an ESLint run returns `1`, there were linting errors. This will be reported anyway, so let's dismiss the code.
- When an ESLint run returns `2`, there was an execution error, e.g. it couldn't write the report to the output file. These are errors that we want to surface, so we'll propagate them through xargs.

mcsf committed:
Avoid single quotes inside Bash script inside Bash script

I initially tried escaping these (\'), but something weird is happening
with TeamCity. In the build we see that this comment inside the quoted
string:

   # only catch relevant issues. ESLint\'s exit codes seem to be:

was echoed in the build log as:

   # only catch relevant issues. ESLint\s' exit codes seem to be:

which breaks the whole syntax. So I'll avoid quotes altogether.

oandregal commented on .teamcity/_self/projects/WebApp.kt:
```diff
@@ -571,23 +571,84 @@ object CheckCodeStyleBranch : BuildType({
+							# only catch relevant issues. ESLint exit codes seem to be:
+							# - 0 for no errors
+							# - 1 for linting errors (should ignore)
+							# - 2 for other errors (should propagate)
```
I tested the `2` code was dispatched by asking eslint to write to a local folder with read-only permissions. It failed as expected. It matches the docs: https://eslint.org/docs/latest/use/command-line-interface#exit-codes

oandregal commented:
Code wise, this looks good. I've rebased this PR to latest `trunk` and updated the test PRs accordingly. If everything goes well, as it went before, I'll go ahead and merge.

oandregal commented:
ok, this is working as expected. It detected an issue in the `packages/components/calendar/README.md` file introduced by https://github.com/Automattic/wp-calypso/pull/101300 I'm fixing it at https://github.com/Automattic/wp-calypso/pull/101309

tyxla commented:
> Code wise, this looks good. I've rebased this PR to latest `trunk` and updated the test PRs accordingly. If everything goes well, as it went before, I'll go ahead and merge.

Thanks for the great work on this very unmaintained area of our CI @oandregal @mcsf!
